### PR TITLE
Fix Edge-case: Mapping fails if the extension code returns nil values #804

### PR DESCRIPTION
**Pull request by inancgumus**
The error output is similar to #820.

Please let me know if you see other edge cases, so I can amend them.

Here are some test snippets you can try.

<details>
<summary>evaluateHandle</summary>

Output:

```
ERRO[0001] Uncaught (in promise) GoError: evaluating handle: result is nil
        at github.com/grafana/xk6-browser/browser.mapPage.func2 (native)
        at file:///Users/inanc/grafana/k6browser/tests/evaluate_handle.js:20:54(31)  executor=per-vu-iterations scenario=default
```

Script:

```js
await page.goto('https://test.k6.io/', { waitUntil: 'load' });
const pageTitleHandle = await page.evaluateHandle(() => {
  return document.querySelector('nothing');
});
```

</details>

<details>
<summary>waitForSelector</summary>

Output:

```
ERRO[0001] Uncaught (in promise) GoError: querying selector "x"
        at github.com/grafana/xk6-browser/browser.mapPage.func13 (native)
        at file:///Users/inanc/grafana/k6browser/tests/waitforselector.js:20:23(32)  executor=per-vu-iterations scenario=default
```

Script:

```js
await page.goto('https://test.k6.io/', { waitUntil: 'load' });
const a = await page.waitForSelector('x');
```

</details>

Closes: https://github.com/grafana/xk6-browser/issues/804
Related: https://github.com/grafana/xk6-browser/issues/688 #820

### CODE REVIEW

File: common/element_handle.go [hunk 1]
```diff
@@ -755,20 +755,21 @@ func (h *ElementHandle) Click(opts goja.Value) error {
...
 		err  error
 	)
 	action := dom.DescribeNode().WithObjectID(h.remoteObject.ObjectID)
 	if node, err = action.Do(cdp.WithExecutor(h.ctx, h.session)); err != nil {
-		k6ext.Panic(h.ctx, "getting remote node %q: %w", h.remoteObject.ObjectID, err)
+		return nil, fmt.Errorf("getting remote node %q: %w", h.remoteObject.ObjectID, err)
 	}
 	if node == nil || node.FrameID == "" {
-		return nil
+		return nil, fmt.Errorf("element is not an iframe")
```
Comment (inancgumus) [inline common/element_handle.go:774]: TODO: Better error message is needed.

File: common/execution_context.go [hunk 2]
```diff
@@ -319,8 +319,12 @@ func (e *ExecutionContext) EvalHandle(
 	}
 	res, err := e.eval(apiCtx, opts, js.ToString().String(), evalArgs...)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("evaluating handle: %w", err)
+	}
+	if res == nil {
```
Comment (ankur22) [inline common/execution_context.go:324]: if we ran:

```javascript
function() {
  console.log("hello world)
}
```

This wouldn't return a result, so would `res` be `nil` in that case?

Is it possible to create a test for this scenario?

File: common/frame.go [hunk 3]
```diff
@@ -758,11 +758,13 @@ func (f *Frame) EvaluateHandle(pageFunc goja.Value, args ...goja.Value) (handle
 	}
 	f.executionContextMu.RUnlock()
 	if err != nil {
-		k6ext.Panic(f.ctx, "evaluating handle: %w", err)
+		// EvalHandle has a rich error, nolinting.
+		return nil, err //nolint:wrapcheck
```
Comment (ankur22) [inline common/frame.go:767]: IMO we should wrap this error with the method name to give the error more context here, instead of wrapping it in `EvalHandle`.

If another method calls `EvalHandle` then `evaluating handle:` might be the wrong context for that caller.

File: common/js_handle.go [hunk 4]
```diff
@@ -101,14 +101,10 @@ func (h *BaseJSHandle) Evaluate(pageFunc goja.Value, args ...goja.Value) any {
...
-func (h *BaseJSHandle) EvaluateHandle(pageFunc goja.Value, args ...goja.Value) api.JSHandle {
+func (h *BaseJSHandle) EvaluateHandle(pageFunc goja.Value, args ...goja.Value) (api.JSHandle, error) {
 	rt := h.execCtx.vu.Runtime()
 	args = append([]goja.Value{rt.ToValue(h)}, args...)
-	res, err := h.execCtx.EvalHandle(h.ctx, pageFunc, args...)
-	if err != nil {
-		k6ext.Panic(h.ctx, "%w", err)
-	}
-	return res
+	return h.execCtx.EvalHandle(h.ctx, pageFunc, args...)
```
Comment (ankur22) [inline common/js_handle.go:107]: Same here, I think we should wrap the error with more context here.

File: common/worker.go [hunk 5]
```diff
@@ -69,9 +69,9 @@ func (w *Worker) Evaluate(pageFunc goja.Value, args ...goja.Value) any {
 }
 
 // EvaluateHandle evaluates a page function in the context of the web worker and returns a JS handle.
-func (w *Worker) EvaluateHandle(pageFunc goja.Value, args ...goja.Value) api.JSHandle {
+func (w *Worker) EvaluateHandle(pageFunc goja.Value, args ...goja.Value) (api.JSHandle, error) {
 	// TODO: implement
-	return nil
+	return nil, nil
```
Comment (ankur22) [inline common/worker.go:74]: Could we use `ErrFatal` instead of `nil` here?

File: browser/mapping.go [hunk 6]
```diff
@@ -248,10 +247,12 @@ func mapElementHandle(vu moduleVU, eh api.ElementHandle) mapping {
 		"type":                   eh.Type,
 		"uncheck":                eh.Uncheck,
 		"waitForElementState":    eh.WaitForElementState,
-		"waitForSelector": func(selector string, opts goja.Value) *goja.Object {
-			eh := eh.WaitForSelector(selector, opts)
-			ehm := mapElementHandle(vu, eh)
-			return rt.ToValue(ehm).ToObject(rt)
```
Comment (ankur22) [inline browser/mapping.go:245]: Why don't we need to use `rt.ToValue(ehm).ToObject(rt)`?

**Review by ankur22 — APPROVED** [review]
LGTM. I left some minor comments, so up to you whether you need to action them or not.

File: common/frame.go (hunk 3 above)
Comment (inancgumus) [reply common/frame.go:767]: You mean something like `EvaluateHandle: evaluating handle:`?

File: common/js_handle.go (hunk 4 above)
Comment (inancgumus) [reply common/js_handle.go:107]: `BaseJSHandle` is an interim method. I don't know what we should add here to enhance the returned error. Would you happen to have a suggestion?

File: common/worker.go (hunk 5 above)
Comment (inancgumus) [reply common/worker.go:74]: Makes sense 👍

File: browser/mapping.go (hunk 6 above)
Comment (inancgumus) [reply browser/mapping.go:245]: Goja handles it automatically. We don't need to return a `goja.Object` or `value`. The current mapping layer implementation unnecessarily uses objects.

File: common/execution_context.go (hunk 2 above)
Comment (inancgumus) [reply common/execution_context.go:324]: Yes, `res` would be `nil` in that case. It's expected since `EvaluateHandle` should return a DOM handle. What would it return for that function?

Comment (inancgumus) [reply common/execution_context.go:324]: Here's a test: https://github.com/grafana/xk6-browser/pull/826/commits/c78b301967870a6033636b4bfe7d9e607ea8f78d

File: common/worker.go (hunk 5 above)
Comment (inancgumus) [reply common/worker.go]: Fatalized here: https://github.com/grafana/xk6-browser/pull/826/commits/04b2e667d203a5b95cf097fc087b480548dcd0e2

File: common/element_handle.go [hunk 7]
```diff
@@ -982,28 +982,31 @@ func (h *ElementHandle) OwnerFrame() api.Frame {
 	}
 	res, err := h.evalWithScript(h.ctx, opts, fn)
 	if err != nil {
-		k6ext.Panic(h.ctx, "getting document element: %w", err)
+		err = fmt.Errorf("getting document element: %w", err)
 	}
-	if res == nil {
-		return nil
+	if res == nil || err != nil {
+		return nil, err
```
Comment (ka3de) [inline common/element_handle.go:993]: I find this part a little bit confusing. Would it be better to be more explicit, such as:
```go
        res, err := h.evalWithScript(h.ctx, opts, fn)
	if err != nil {
		return nil, fmt.Errorf("getting document element: %w", err)
	}
	if res == nil {
		return nil, nil
	}
```
?

**Review by ka3de — APPROVED** [review]
Thanks for working on this! Seems like we still have things to learn about Goja :smile: 
Made a small comment, but overall LGTM.

File: common/element_handle.go (hunk 7 above)
Comment (inancgumus) [reply common/element_handle.go]: > Thanks for working on this! Seems like we still have things to learn about Goja 😄

Yeah 😄 Thanks!

> I find this part a little bit confusing.

Explicitified here: https://github.com/grafana/xk6-browser/pull/826/commits/d4aec1fafd7cb909d761a5ba54bb54b5339214ee

File: common/frame.go (hunk 3 above)
Comment (ankur22) [reply common/frame.go:767]: I agree that `EvaluateHandle: evaluating handle:` is not a good approach.

This is very minor so, don't worry about changing the code if it doesn't make sense. What i meant was that instead of wrapping the error in `EvalHandle` like so:

```go
return nil, fmt.Errorf("evaluating handle: %w", err)
```

why not just return `err` and allow this method (`EvaluateHandle`) to wrap the error with:

```go
 fmt.Errorf("evaluating handle: %w", err)
```

Another example is, in `EvalHandle` when `res` is `nil` it is currently wrapped like so:

```go
return nil, fmt.Errorf("evaluating handle: result is nil")
```

When it could be:

```go
return nil, fmt.Errorf("result is nil")
```

And in `EvaluateHandle` we would wrap the `err` with:

```go
fmt.Errorf("evaluating handle: %w", err).
```

Like i said, super minor so don't worry about this 🙂

File: common/js_handle.go (hunk 4 above)
Comment (ankur22) [reply common/js_handle.go:107]: Ah, ok, that's fine in that case. I wouldn't worry about changing anything here. It was minor anyway 🙂

**Review by ankur22 — APPROVED** [review]

File: common/frame.go (hunk 3 above)
Comment (inancgumus) [reply common/frame.go]: @ankur22, oh, I see what you mean now 🙂 Thanks for explaining 🙇

I couldn't see a difference in adding `"evaluating handle"` in `EvalHandle` vs. in`Page|Frame|JSHandle (ElementHandle)` since the final error message would be the same. So I've decided to go this way: [f040f90](https://github.com/grafana/xk6-browser/pull/826/commits/f040f9019edaeca5e017cbe577742d1059cee3de) to differentiate the error messages depending on where `EvalHandle` is called. Please check it out and let me know; thanks!

Comment (ankur22) [reply common/frame.go]: 🎉 Nice! That's exactly what i was thinking. Thanks 🙂

**Review by ankur22 — APPROVED** [review]
